packages/proto-rpc/src/ServerRegistry.ts

```diff
@@ -0,0 +1,134 @@
+import { RpcMessage } from './proto/ProtoRpc'
+import EventEmitter = require('events')
+import { MethodInfo, RpcMetadata, RpcStatus, ServerCallContext } from '@protobuf-ts/runtime-rpc'
+import { BinaryReadOptions, BinaryWriteOptions } from '@protobuf-ts/runtime'
+import { promiseTimeout } from './common'
+import { Err } from './errors'
+import UnknownRpcMethod = Err.UnknownRpcMethod
+import { Logger } from './Logger'
+import { ConversionWrappers } from './ConversionWrappers'
+import { ProtoRpcOptions } from './ClientTransport'
+import { Empty } from './proto/google/protobuf/empty'
+
+export enum ServerRegistryEvent {
+    RPC_RESPONSE = 'streamr:dht-transport:server:response-new',
+    RPC_REQUEST = 'streamr:dht-transport:server:request-new',
+}
+
+export interface ServerRegistry {
+    on(event: ServerRegistryEvent.RPC_RESPONSE, listener: (rpcMessage: RpcMessage) => void): this
+    on(event: ServerRegistryEvent.RPC_REQUEST, listener: (rpcMessage: RpcMessage) => void): this
+}
+
+export interface Parser<Target> { fromBinary: (data: Uint8Array, options?: Partial<BinaryReadOptions>) => Target }
+export interface Serializer<Target> { toBinary: (message: Target, options?: Partial<BinaryWriteOptions>) => Uint8Array }
+
+export type RegisteredMethod = (request: Uint8Array, callContext: CallContext) => Promise<Uint8Array>
+export type RegisteredNotification = (request: Uint8Array, callContext: CallContext) => Promise<Empty>
+
+const logger = new Logger(module)
+
+export class ServerRegistry extends EventEmitter {
+    methods: Map<string, RegisteredMethod | RegisteredNotification>
+    private stopped = false
+    constructor() {
+        super()
+        this.methods = new Map()
+    }
+
+    public async onRequest(rpcMessage: RpcMessage, callContext?: CallContext): Promise<Uint8Array> {
+        if (this.stopped) {
+            return new Uint8Array()
+        }
+        logger.trace(`Server processing request ${rpcMessage.requestId}`)
+        const methodName = rpcMessage.header.method
+        const fn = this.methods.get(methodName) as RegisteredMethod
+        if (!fn) {
+            throw new UnknownRpcMethod(`RPC Method ${methodName} is not provided`)
+        }
+
+        return await promiseTimeout(1000, fn!(rpcMessage.body, callContext ? callContext : new CallContext()))
+    }
+
+    public async onNotification(rpcMessage: RpcMessage, callContext?: CallContext): Promise<Empty> {
+        if (this.stopped) {
+            return {} as Empty
+        }
+        logger.trace(`Server processing notification ${rpcMessage.requestId}`)
+        const methodName = rpcMessage.header.method
+        const fn = this.methods.get(methodName) as RegisteredNotification
+        if (!fn) {
+            throw new UnknownRpcMethod(`RPC Method ${methodName} is not provided`)
+        }
+        return await promiseTimeout(1000, fn!(rpcMessage.body, callContext ? callContext : new CallContext()))
+    }
```

Comment:
```suggestion
    public async onNotification(rpcMessage: RpcMessage, callContext?: CallContext): Promise<void> {
        if (this.stopped) {
            return
        }
        logger.trace(`Server processing notification ${rpcMessage.requestId}`)
        const methodName = rpcMessage.header.method
        const fn = this.methods.get(methodName) as RegisteredNotification
        if (!fn) {
            throw new UnknownRpcMethod(`RPC Method ${methodName} is not provided`)
        }
        await promiseTimeout(1000, fn!(rpcMessage.body, callContext ? callContext : new CallContext()))
    }
```

Here also we should check that `methodName !== undefined` and that the returned `fn` is actually a `RegisteredNotification`.

Reply:
Refactored by splitting RPC calls and notifications in separate Maps